sdk/schemaregistry/Microsoft.Azure.Data.SchemaRegistry.ApacheAvro/src/SchemaRegistryAvroEncoder.cs

```diff
@@ -217,28 +211,39 @@ private static DatumReader<object> GetReader(Schema writerSchema, Schema readerS
         /// <param name="messageFactory">Optional func to create a derived instance of <see cref="MessageWithMetadata"/> given the serialized Avro.
         /// If not specified, it is assumed that the derived type has a public constructor accepting a <see cref="BinaryData"/> instance.</param>
         /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        /// <typeparam name="T">The <see cref="MessageWithMetadata"/> type to encode the data into.</typeparam>
         public async ValueTask<T> EncodeMessageDataAsync<T>(
             object data,
             Type inputType = default,
             Func<BinaryData, T> messageFactory = default,
             CancellationToken cancellationToken = default) where T : MessageWithMetadata
             => await EncodeMessageDataInternalAsync(data, inputType, messageFactory, true, cancellationToken).ConfigureAwait(false);
 
-        private async ValueTask<T> EncodeMessageDataInternalAsync<T>(
+        internal async ValueTask<T> EncodeMessageDataInternalAsync<T>(
             object data,
             Type inputType,
             Func<BinaryData, T> messageFactory,
             bool async,
             CancellationToken cancellationToken) where T : MessageWithMetadata
         {
             (string schemaId, BinaryData bd) = async
-                ? await EncodeAsync(data, inputType ?? data?.GetType(), cancellationToken).ConfigureAwait(false)
-                : Encode(data, inputType ?? data?.GetType(), cancellationToken);
+                ? await EncodeInternalAsync(data, inputType, true, cancellationToken).ConfigureAwait(false)
+                : EncodeInternalAsync(data, inputType, false, cancellationToken).EnsureCompleted();
 
             MessageWithMetadata message;
             if (messageFactory == default)
             {
-                message = (MessageWithMetadata)Activator.CreateInstance(typeof(T), bd);
+                if (typeof(T) == typeof(MessageWithMetadata))
+                {
+                    // If concrete type is used, we need to use the parameterless constructor as the concrete type does
+                    // not have any other constructors by design (to make it easier to use across the different messaging libraries).
+                    message = (MessageWithMetadata)Activator.CreateInstance(typeof(T));
+                    message.Data = bd;
+                }
+                else
+                {
+                    message = (MessageWithMetadata)Activator.CreateInstance(typeof(T), bd);
```

Comment:
How do we know the type will implement a ctor accepting `BinaryData`?

Reply:
This is essentially a convention that implementing types will follow. For types that don't adhere to this, a messageFactory needs to be passed.

Reply:
We could possibly improve the error scenario though.

Reply:
I'm not familiar with this type, so I'm just curious how an implementor would know this? Should the `MessageWithMetadata` implement it too and just set Data to it?

Reply:
In general, other messaging libraries are the implementors - e.g. Event Hubs, Service Bus, and Event Grid.

I was going back and forth on adding the ctor to MessageWithMetadata, but I think you are right - we can just add it. If an inheriting type overrides the property, then the property from the base type will never be used. If they don't override it, then they will have the Data populated.

Reply:
Actually.. I don't think this is correct - adding this constructor to the base type won't help ensure that child types have the same constructor, so I don't think there is a benefit to adding the constructor.